### DISCUSSION TITLE
RESTEasy Reactive REST client causes SocketException minutes after AWS lambda has finished

### DISCUSSION THREAD

**Discussion by gefloh** [quarkusio/quarkus#30824] [Q&A] [answered]
Hello,

currently I am migrating several components from RESTEasy Classic to Reactive (just switching the dependency at first, not changing the way REST is used). In one project which is an AWS lambda and doesn't have own REST resources but uses the reactive REST client, I am seeing a strange behavior which causes me some headache. After roughly 6 minutes and after the lambda has long finished running, Quarkus seems to continue using the former connection which causes then a SocketException. Enabling detailed logs for the REST client shows:

```
2023-02-02T14:59:33.362+01:00 Duration: x ms	Billed Duration: x ms	Memory Size: x MB	Max Memory Used: x MB	Init Duration: x ms	
2023-02-02T15:05:20.705+01:00 [id: 0x19ba1ce9, L:/<ip>:48654 - R:<host>/<ip>:443] READ COMPLETE
2023-02-02T15:05:20.705+01:00 [id: 0x19ba1ce9, L:/<ip>:48654 - R:<host>/<ip>:443] EXCEPTION: java.net.SocketException: Connection reset
2023-02-02T15:05:20.705+01:00 [id: 0x19ba1ce9, L:/<ip>:48654 - R:<host>/<ip>:443] CLOSE
2023-02-02T15:05:20.705+01:00 [id: 0x19ba1ce9, L:/<ip>:48654 ! R:<host>/<ip>:443] USER_EVENT: io.netty.channel.socket.ChannelInputShutdownReadComplete@2bb1f694
2023-02-02T15:05:20.705+01:00 [id: 0x19ba1ce9, L:/<ip>:48654 ! R:<host>/<ip>:443] USER_EVENT: SslCloseCompletionEvent(java.nio.channels.ClosedChannelException)
2023-02-02T15:05:20.705+01:00 [id: 0x19ba1ce9, L:/<ip>:48654 ! R:<host>/<ip>:443] INACTIVE
2023-02-02T15:05:20.705+01:00 [id: 0x19ba1ce9, L:/<ip>:48654 ! R:<host>/<ip>:443] UNREGISTERED
```

This seems to happen on the I/O thread:

    "threadName": "vert.x-eventloop-thread-1"

Does anyone know what is causing this? What can I do to avoid this?

Thanks,
Gerhard

[a comment by an automated account is not shown]

**Reply by cescoffier** [reply]
If I understand correctly, you have a lambda using the REST client to call another service. Is it that connection that stays open? 
The pool keeps the connection open (so subsequent calls will be faster). You can also disable the "keep-alive" option to force close the connection.

**Reply by gefloh** [reply]
Thanks for your reply. 

> If I understand correctly, you have a lambda using the REST client to call another service. Is it that connection that stays open?

Yes, exactly that.

How would I do that? `quarkus.rest-client.connection-ttl`?

Interesting thing is that with the RESTEasy Classic REST client this problem has not occurred (or at least it didn't show up in the logs).

**Reply by geoand** [reply]
Yes, `quarkus.rest-client.connection-ttl` is the property that corresponds to the underlying Vert.x `HttpClientOptions#keepAliveTimeout`

**Reply by gefloh** [reply]
I played a bit around with this parameter, still the situation stays unclear for me. The exception still occurs, even with a value of e.g. 1000 (ms) it is not gone and still occurs roughly 6 minutes (sometimes less) after the lambda has finished which makes not much sense to me. 
Additionally unfortunately I see three different scenarios:

- The one mentioned above where after a while the exception occurs
- Sometimes the lambda just finishes as expected with nothing happening later
- Sometimes we have

```
03.02.2023 13:35:24.012 [id: 0x40792513, L:/<ip>:46942 - R:<host>/<ip>:443] WRITE: 0B
03.02.2023 13:35:24.012 [id: 0x40792513, L:/<ip>:46942 - R:<host>/<ip>:443] FLUSH
03.02.2023 13:35:24.012 [id: 0x40792513, L:/<ip>:46942 - R:<host>/<ip>:443] CLOSE
03.02.2023 13:35:24.014 [id: 0x40792513, L:/<ip>:46942 ! R:<host>/<ip>:443] USER_EVENT: SslCloseCompletionEvent(java.nio.channels.ClosedChannelException)
03.02.2023 13:35:24.014 [id: 0x40792513, L:/<ip>:46942 ! R:<host>/<ip>:443] INACTIVE
03.02.2023 13:35:24.014 [id: 0x40792513, L:/<ip>:46942 ! R:<host>/<ip>:443] UNREGISTERED
```
after the lambda has finished which also looks "normal".

I tried increasing the log level to TRACE but that didn't provide further information. I also tried to set `quarkus.rest-client.connection-pool-size` to 1, no success. Any other ideas?

**Reply by gefloh** [reply]
After a bunch of more experiments I have at least found a workaround in using a custom `HttpClientOptions` and disabling the keep-alive completely

    final HttpClientOptions options = new HttpClientOptions();
    options.setKeepAlive(false);

With this setting, the exception seems to be gone and from the performance side it is not an issue in this case.

Remaining unclear:

- What causes the initial problem?
- Why does the exception occur roughly after 6 minutes? All timeouts I have found are far shorter.

**Reply by gefloh** [reply]
Guys, I would like to come back to this once again with more info/questions. This timeout of about 6 minutes we see comes from AWS shutting down the Lambda container, so this timeout is not Quarkus/Vert.x related. Further observations: The problem only exists when the Lambda is compiled natively, non-natively compiled everything is smooth. We see it in all our native compiled Lambdas, so it's not a specific thing of one Lambda. However, this means that we have to turn off connection pooling for all Lambdas which use RESTEasy Reactive natively. It seems that somethings goes wrong during shutdown phase when compiled natively.

Further questions, @geoand you're the expert ;-) :
- Can we somehow influence the shutdown behavior of a natively compiled Quarkus Lambda? The AWS support said that we might have to do some manual cleanup but I don't know how this could be done?
- Any idea how we could simulate this locally? I tried it with SAM CLI  (there is even a `--shutdown` option to simulate the shutdown) but without success.

Thanks!

**Reply by geoand** [reply]
@galderz @patriot1burke do you have any idea why the shutdown handlers would not work properly for AWS Lambda in native mode?

**Reply by galderz** [reply]
@geoand Well, the implementation is completely different due to substrate VM so maybe there's a bug in there. Nothing rings a bell right now. @gefloh can you provide a reproducer that we can run? Or if someone understands how AWS does shutdown, something can be emulated?

**Reply by gefloh** [reply]
@galderz Well, I probably can create a small Lambda for reproducing that. The thing is: I couldn't simulate this problem locally yet, which means, you would have to upload it to AWS to run it. I am not sure if that helps then.

**Reply by stuartwdouglas** [reply]
So the exception is basically on the client side, when AWS is shutting down the client lambda container? 

I would guess there is some difference in how the close is being reported, e.g. maybe in JVM mode the socket just returns -1 while in native it throws an exception.

**Reply by gefloh** [reply]
Yes, although in the meantime I tend to give up on this and go with the workaround of disabling the keep alive completely. I tried to create a reproducer but found that with a basic Lambda everything is working perfectly. Then I added different dependencies that are used by our Lambdas and the error was there, so it seems one of the dependencies causes this behavior.
Still, searching for the problem is extremely time consuming because:

- Native compilation is needed, which takes times
- I have to run the Lambda and wait for the shutdown through AWS which takes several minutes
- AND the issue does not occur on every run, sometimes the Lambda has to be run several times, so probably it's also an issue with the async nature of Vert.x. 

Currently it looks more reasonable to use the workaround and try later again with the keep alive when "something" (tm) has changed in the dependencies.

**Comment by gefloh** [comment, marked as answer]
Workaround to avoid the exception (see discussion above):

    final HttpClientOptions options = new HttpClientOptions();
    options.setKeepAlive(false);

Thanks @cescoffier for the hint with keep-alive, thanks @geoand.

**Comment by geoand** [comment]
Interesting enough, we don't see to have a way to turn off  Vertx's keep-alive for the Reactive REST Client... @TomasHofman would you like to add that?

**Reply by TomasHofman** [reply]
I will take a look, but it will probably take me a week or two before I have time. So if anybody wants to do it sooner that's fine ;).

**Reply by geoand** [reply]
No rush 🙂

**Reply by gefloh** [reply]
I could also have a look and add this. It doesn't sound too complicated. ;-)

**Reply by TomasHofman** [reply]
Sure, go for it.
